File: solutions/lua/triangle/1/triangle.lua
local triangle = {}

function triangle.kind(a, b, c)
    if a <= 0 or b <= 0 or c <= 0 then
        error('Input Error')
    end

    if a + b < c or b + c < a or a + c < b then
        error('Input Error')
    end
    
    local type = ''
    if ( a == b and a == c ) then
        type = 'equilateral'
    elseif ( a ~= b and a ~= c and b ~= c ) then
        type = 'scalene'
    else
        type = 'isosceles'
    end
    return type
end

return triangle

File: solutions/prolog/triangle/1/triangle.pl
triangle(Side1, Side2, Side3, Type) :-
    (
        S is Side1 + Side2, S < Side3      -> false;
        S is Side1 + Side3, S < Side2      -> false;
        S is Side2 + Side3, S < Side1      -> false;
        Side1 =< 0, Side2 =< 0, Side3 =< 0 -> false;
        Side1  == Side2, Side2  == Side3, Type == "equilateral" -> true;
        Side1  == Side2, Side2  == Side3, Type == "isosceles"   -> true;
        Side1 \== Side2, Side2 \== Side3, Type == "scalene"     -> true;
        Side1  == Side2, Side1 \== Side3, Type == "isosceles"   -> true;
        Side1  == Side3, Side2 \== Side3, Type == "isosceles"   -> true;
        Side2  == Side3, Side1 \== Side2, Type == "isosceles"   -> true;
        false
    )
.
